File: torch-sys-cuda/src/bridge_rocm.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

#[allow(dead_code)]
#[cxx::bridge(namespace = "monarch")]
pub(crate) mod ffi {
    unsafe extern "C++" {
        include!("ATen/hip/HIPEvent.h");
        include!("monarch/torch-sys-cuda/src/bridge_rocm.h");

        // HIP Event APIs (uses at::cuda namespace for compatibility)
        #[namespace = "at::cuda"]
        type CUDAEvent;
        fn create_hip_event(
            enable_timing: bool,
            blocking: bool,
            interprocess: bool,
        ) -> UniquePtr<CUDAEvent>;
        fn record_event(event: Pin<&mut CUDAEvent>, stream: &HIPStream);
        fn block_event(event: Pin<&mut CUDAEvent>, stream: &HIPStream);
        fn query(self: &CUDAEvent) -> bool;
        fn elapsed_time(self: &CUDAEvent, end_event: &CUDAEvent) -> f32;
        fn synchronize(self: &CUDAEvent);

        // HIP Stream APIs
        #[namespace = "c10::hip"]
        type HIPStream;
        fn get_current_hip_stream(device: i8) -> SharedPtr<HIPStream>;
        fn set_current_hip_stream(stream: &HIPStream);
        fn create_hip_stream(device: i8, priority: i32) -> SharedPtr<HIPStream>;
        fn query(self: &HIPStream) -> bool;
        fn synchronize(self: &HIPStream);
        fn device_index(self: &HIPStream) -> i8;
        // Returns the raw HIP stream handle as usize
        fn get_stream_handle(stream: &HIPStream) -> usize;

        // RCCL helpers
        #[namespace = ""]
        type ncclConfig_t = nccl_sys::ncclConfig_t;
        fn make_nccl_config() -> ncclConfig_t;
    }
}

use std::fmt::Debug;
use std::fmt::Error;
use std::fmt::Formatter;

// SAFETY: HIPStream is thread safe
unsafe impl Send for ffi::HIPStream {}
// SAFETY: see above
unsafe impl Sync for ffi::HIPStream {}

impl Debug for ffi::HIPStream {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
        f.debug_struct("HIPStream")
            .field("device", &format!("{}", self.device_index()))
            .field("stream", &format!("{:p}", self))
            .finish()
    }
}

impl Debug for ffi::CUDAEvent {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
        f.debug_struct("CUDAEvent")
            .field("ptr", &format!("{:p}", self))
            .finish()
    }
}

// SAFETY: CUDAEvent is thread safe
unsafe impl Send for ffi::CUDAEvent {}
// SAFETY: see above
unsafe impl Sync for ffi::CUDAEvent {}

File: torch-sys-cuda/src/bridge_rocm.cpp
#include "monarch/torch-sys-cuda/src/bridge_rocm.h"

namespace monarch {

std::unique_ptr<at::cuda::CUDAEvent>
create_hip_event(bool enable_timing, bool blocking, bool interprocess) {
  unsigned int flags = (blocking ? hipEventBlockingSync : hipEventDefault) |
      (enable_timing ? hipEventDefault : hipEventDisableTiming) |
      (interprocess ? hipEventInterprocess : hipEventDefault);

  return std::make_unique<at::cuda::CUDAEvent>(flags);
}

void record_event(at::cuda::CUDAEvent& event, const c10::hip::HIPStream& stream) {
  at::hip::HIPStreamMasqueradingAsCUDA masq_stream(stream);
  event.record(masq_stream);
}

void block_event(at::cuda::CUDAEvent& event, const c10::hip::HIPStream& stream) {
  at::hip::HIPStreamMasqueradingAsCUDA masq_stream(stream);
  event.block(masq_stream);
}

std::shared_ptr<c10::hip::HIPStream> get_current_hip_stream(
    c10::DeviceIndex device) {
  return std::make_shared<c10::hip::HIPStream>(
      c10::hip::getCurrentHIPStream(device));
}

std::shared_ptr<c10::hip::HIPStream> create_hip_stream(
    c10::DeviceIndex device,
    int32_t priority) {
  return std::make_shared<c10::hip::HIPStream>(
      c10::hip::getStreamFromPool((const int)priority, device));
}

void set_current_hip_stream(const c10::hip::HIPStream& stream) {
  auto device = c10::hip::current_device();
  if (device != stream.device_index()) {
    c10::hip::set_device(stream.device_index());
  }
  at::hip::setCurrentHIPStream(stream);
}

size_t get_stream_handle(const c10::hip::HIPStream& stream) {
  return reinterpret_cast<size_t>(stream.stream());
}

} // namespace monarch
